### PR TITLE
new bucket for tables, master kms and, iam roles for groups

### PR DESCRIPTION
**Pull request by MattToczek**
* New s3 for DBs
* Policies for access based on group 
* HDFS setup has hive db setup added

### CODE REVIEW

File: terraform/modules/emr/iam.tf [hunk 1]
```diff
@@ -573,3 +590,71 @@ data "aws_iam_policy_document" "elastic_map_reduce_for_auto_scaling_role" {
     resources = ["*"] // Required by AutoScaling-CheckPermissions 
   }
 }
+
+data "aws_iam_policy_document" "group_hive_data_access_documents" {
+  for_each = var.security_configuration_groups
+
+  policy_id = "HiveData${replace(regex("((Non_)?PII)", each.key)[0], "_", "")}"
```
Comment (IanUtley) [inline terraform/modules/emr/iam.tf:597]: Whats this Non_ PII references?  Why are we hardcoding parts of group names? There will be new groups added in future, what does that mean for this regex?

File: terraform/modules/emr/iam.tf [hunk 2]
```diff
@@ -573,3 +590,71 @@ data "aws_iam_policy_document" "elastic_map_reduce_for_auto_scaling_role" {
...
+  }
+
+  statement {
+    sid = "HiveDataS3"
+    actions = [
+      "s3:GetBucketPublicAccessBlock",
+      "s3:ListBucket",
+      "s3:GetBucketVersioning",
+      "s3:ListMultipartUploadParts",
+      "s3:GetObject",
```
Comment (IanUtley) [inline terraform/modules/emr/iam.tf:625]: Doesnt this let me read the data from another groups database? ie. non PII user able to read tables in the PII database?

File: terraform/modules/emr/iam.tf [hunk 3]
```diff
@@ -573,3 +590,71 @@ data "aws_iam_policy_document" "elastic_map_reduce_for_auto_scaling_role" {
...
+      "s3:GetBucketLocation"
+    ]
+    resources = [
+      "${aws_s3_bucket.hive_data.arn}/*",
+      "${aws_s3_bucket.hive_data.arn}/"
+    ]
+  }
+
+  statement {
+    sid = "HiveDataKms${replace(regex("((Non_)?PII)", each.key)[0], "_", "")}"
```
Comment (IanUtley) [inline terraform/modules/emr/iam.tf:635]: See regex comment above

File: terraform/modules/emr/iam.tf (hunk 1 above)
Comment (MattToczek) [reply terraform/modules/emr/iam.tf:597]: Sorry - I meant to amend the regex to be similar to [this sed](https://github.com/dwp/aws-analytical-env/pull/193/files/5883cc6606e6513180f053e2c00d6fe4471ed3e8#diff-301f04c512e663a7e7389eea0d85f032facb55a7ad5e812f86cefcccceb3713aR42) to catch any non-alphanumeric chars but missed it when I raised the PR. Will sort now.

File: terraform/modules/emr/templates/emr/hdfs_setup.sh [hunk 4]
```diff
@@ -35,3 +37,8 @@ sudo systemctl start hive-server2
 sudo -H -u hdfs bash -c "hdfs dfs -mkdir /libs"
 sudo -H -u hdfs bash -c "hdfs dfs -put /usr/lib/hbase/hbase-client.jar /libs"
 
+for EACH_GROUP in $${USER_GROUPS[@]}; do
+  echo "Creating group DB in S3 for '$EACH_GROUP'"
+  str=$(echo $EACH_GROUP | sed -e 's/[^a-zA-Z0-9_]/_/g')
+  sudo /bin/hive -e "CREATE DATABASE IF NOT EXISTS $${str}_db LOCATION '$S3_FILE_PATH/$str/$${str}_db'"
```
Comment (IanUtley) [inline terraform/modules/emr/templates/emr/hdfs_setup.sh:43]: Whats with the double $$? The str wont be env var, so ${str} so won't $${str}_db always resolve to '_db'?

Comment (IanUtley) [reply terraform/modules/emr/templates/emr/hdfs_setup.sh:43]: Why we using $$?  Looks like the previous version worked fine?

Comment (MattToczek) [reply terraform/modules/emr/templates/emr/hdfs_setup.sh:43]: It is a template file so tf is looking for `${<value>}` to template into it - the `$$` means that tf leaves it alone and it is able to plan/apply without complaining about missing template vars.

**Review by IanUtley — APPROVED** [review]

**Pull request merged**